libc/string_catdup.c

```diff
@@ -0,0 +1,689 @@
+}
+
+
+TEST(string_cat, strncat_huge_string_first)
+{
+	char huge_buf[PATH_MAX] = "";
+	char buf[PATH_MAX] = "";
+	char buf1[] = "";
+
+	for (int i = 0; i < PATH_MAX - 1; i++) {
+		huge_buf[i] = HUGE_STRING[i];
+		buf[i] = HUGE_STRING[i];
+	}
+
+	huge_buf[PATH_MAX - 1] = '\0';
+	buf[PATH_MAX - 1] = '\0';
+
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+
+	TEST_ASSERT_EQUAL_STRING(huge_buf, strncat(buf, buf1, 1));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strncat_huge_string_sec)
+{
+	char huge_buf[PATH_MAX] = "";
+	char buf[PATH_MAX + 1] = "";
+	char buf1[PATH_MAX] = "";
+
+	for (int i = 0; i < PATH_MAX - 1; i++) {
+		huge_buf[i] = HUGE_STRING[i];
+		buf1[i] = HUGE_STRING[i];
+	}
+
+	huge_buf[PATH_MAX - 1] = '\0';
+	buf1[PATH_MAX - 1] = '\0';
+
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf1);
+
+	TEST_ASSERT_EQUAL_STRING(huge_buf, strncat(buf, buf1, PATH_MAX));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strncat_two_huge_string)
+{
+	char huge_buf[PATH_MAX] = "";
+	char buf[PATH_MAX] = "";
+	char buf1[PATH_MAX / 2] = "";
+
+	for (int i = 0; i < PATH_MAX - 2; i++) {
+		if (i < PATH_MAX / 2 - 1) {
+			huge_buf[i] = HUGE_STRING[i];
+			buf[i] = HUGE_STRING[i];
+			buf1[i] = HUGE_STRING[i];
+		}
+		else {
+			huge_buf[i] = HUGE_STRING[i - (PATH_MAX / 2 - 1)];
+		}
+	}
+
+	huge_buf[PATH_MAX - 2] = '\0';
+	buf[PATH_MAX / 2 - 1] = '\0';
+	buf[PATH_MAX / 2 - 1] = '\0';
+
+	TEST_ASSERT_EQUAL_STRING(huge_buf, strncat(buf, buf1, PATH_MAX / 2));
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf);
+}
+
+
+TEST(string_cat, strncat_special_char)
+{
+	char buf[SIZE_BUF * 4] = SPECIAL_CHAR;
+	char buf1[] = SPECIAL_CHAR;
+	TEST_ASSERT_EQUAL_STRING(SPECIAL_CHAR SPECIAL_CHAR, strncat(buf, buf1, strlen("Lorem\t !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~ipsum\n !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~")));
+	TEST_ASSERT_EQUAL_STRING(SPECIAL_CHAR SPECIAL_CHAR, buf);
+}
+
+TEST(string_cat, strncat_ascii)
+{
+	int c, count = 0;
+	char buf[512] = { 0 };
+	char ascii[256] = { 0 };
+
+	for (c = 0; c < 255; c++) {
+		if (isprint(c)) {
+			ascii[count++] = c;
+		}
+	}
+	ascii[count++] = '\0';
+
+	TEST_ASSERT_EQUAL_STRING(ascii, strncat(buf, ascii, 256));
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+}
+
+TEST(string_cat, strncat_non_ascii)
+{
+	char buf[SIZE_BUF * 2] = NON_ASCII_1;
+	char buf1[] = NON_ASCII_2;
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, strncat(buf, buf1, strlen(NON_ASCII_1 NON_ASCII_2)));
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, buf);
+}
+
+/////////////////////////////////////////////////////////////////////////
+
+TEST_SETUP(string_dup)
+{
+}
+
+
+TEST_TEAR_DOWN(string_dup)
+{
+}
+
+
+TEST(string_dup, strdup_basic)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_empty)
+{
+	char buf[SIZE_BUF] = "";
```

Comment:
I know that we want to be consistent, but allocating 32 bytes for 2 NUL chars seems unnecessary to me.

```suggestion
	char buf[2] = "";
```